lib/util.c

```diff
@@ -144,24 +144,27 @@ void fprintw(FILE *f, const char *s, size_t w)
 	free(news);
 }
 
-uint32_t parse_max_hosts(char *max_targets)
+// parse a string representing a number/percentage of targets
+// A percentage is interpreted as percent of the search space, so 2^32 * port_count
+uint64_t parse_max_targets(char *max_targets, int port_count)
 {
+	assert(port_count > 0);
 	char *end;
 	errno = 0;
 	double v = strtod(max_targets, &end);
 	if (end == max_targets || errno != 0) {
 		log_fatal("argparse", "can't convert max-targets to a number");
 	}
 	if (end[0] == '%' && end[1] == '\0') {
-		// treat as percentage
-		v = v * ((unsigned long long int)1 << 32) / 100.;
+		// treat as percentage of the search space, so percent of 2^32 * num_ports
+		v = v * (((unsigned long long int)1 << 32) * port_count) / 100.;
```

Comment:
I would define this using something like `uint64_t` to be explicit about what you want here. We do that other places within the code base.